bundles/com.salesforce.bazel.eclipse.core/src/com/salesforce/bazel/eclipse/core/model/discovery/ProjectPerTargetProvisioningStrategy.java

```diff
@@ -252,4 +284,91 @@ protected BazelProject provisionTargetProject(BazelTarget target, IProgressMonit
         return target.getBazelProject();
     }
 
+    /**
+     * Collects all BazelPackages that are referenced by the dependencies of the given projects. This is used for
+     * performance optimization - by pre-loading these packages in batch, we avoid repeated individual Bazel query
+     * commands during classpath resolution.
+     *
+     * @param bazelProjects
+     *            the projects whose dependencies should be analyzed
+     * @param aspectsInfo
+     *            the aspects information containing dependency data
+     * @param workspace
+     *            the Bazel workspace
+     * @return set of BazelPackages to pre-load
+     */
+    private Set<BazelPackage> collectDependencyPackages(Collection<BazelProject> bazelProjects,
+            JavaAspectsInfo aspectsInfo, BazelWorkspace workspace) {
+        Set<BazelPackage> packages = new HashSet<>();
+
+        for (BazelProject bazelProject : bazelProjects) {
+            try {
+                // Skip projects that don't have a proper target yet
+                if (!bazelProject.isTargetProject() || bazelProject.getBazelTarget() == null) {
+                    LOG.trace("Skipping project {} - not a target project or target is null", bazelProject.getName());
+                    continue;
+                }
+
+                // Collect all dependency labels from the aspects info
+                var targetKey = TargetKey.forPlainTarget(bazelProject.getBazelTarget().getLabel().toPrimitive());
+                var targetInfo = aspectsInfo.get(targetKey);
+
+                if (targetInfo != null) {
+                    // Collect packages from direct dependencies
+                    targetInfo.getDependencies().forEach(dep -> {
+                        addPackageForLabel(dep.getTargetKey().getLabel(), workspace, packages);
+                    });
+
+                    // Collect packages from runtime dependencies
+                    var runtimeClasspath = aspectsInfo.getRuntimeClasspath(targetKey);
+                    if (runtimeClasspath != null) {
+                        runtimeClasspath.forEach(jar -> {
```

Comment:
It might be easier to use a classic `forech` loop and inline to allow exceptions to propagate.